unittests/integration/builder_tester.cpp

```diff
@@ -132,6 +132,40 @@ CUDAQ_TEST(BuilderTester, checkSimple) {
   }
 }
 
+CUDAQ_TEST(BuilderTester, checkSwap) {
```

Comment:
Only works on QPP backend

Reply:
Why do you say that? I just pulled this branch and this test passes on qpp and custatevec-fp32 for me. 
```
unittests/test_runtime_custatevec-fp32 --gtest_filter=*BuilderTester.checkSwap
```

Reply:
That's my fault, I meant to tag this PR comment on NVQIRTester.checkSWAP 🤦

Reply:
Those NVQIRTester tests fail because the qubit Arrays are not deallocated. 

```diff
diff --git a/unittests/qir/NVQIRTester.cpp b/unittests/qir/NVQIRTester.cpp
index ad33dba..933a05f 100644
--- a/unittests/qir/NVQIRTester.cpp
+++ b/unittests/qir/NVQIRTester.cpp
@@ -154,7 +154,6 @@ CUDAQ_TEST(NVQIRTester, checkQuantumIntrinsics) {
 
 // FIXME: Disabling test as qpp is the only backend capable
 // of swapping qubits right now.
-#if 0
 CUDAQ_TEST(NVQIRTester, checkSWAP) {
   // Simple SWAP.
   {
@@ -173,6 +172,9 @@ CUDAQ_TEST(NVQIRTester, checkSWAP) {
 
     assert(*__quantum__qis__mz(q0) == 0);
     assert(*__quantum__qis__mz(q1) == 1);
+
+    __quantum__rt__qubit_release_array(qubits);
+    __quantum__rt__finalize();
   }
 
   // SWAP with a single ctrl qubit in 0 state.
@@ -195,6 +197,10 @@ CUDAQ_TEST(NVQIRTester, checkSWAP) {
 
     assert(*__quantum__qis__mz(q0) == 0);
     assert(*__quantum__qis__mz(q1) == 1);
+
+    __quantum__rt__qubit_release_array(ctrls);
+    __quantum__rt__qubit_release_array(qubits);
+    __quantum__rt__finalize();
   }
 
   // SWAP with three ctrl qubits in 1 state.
@@ -226,13 +232,12 @@ CUDAQ_TEST(NVQIRTester, checkSWAP) {
 
     assert(*__quantum__qis__mz(q0) == 1);
     assert(*__quantum__qis__mz(q1) == 0);
+    __quantum__rt__qubit_release_array(ctrls);
+    __quantum__rt__qubit_release_array(qubits);
+    __quantum__rt__finalize();
   }
 }
-#endif
 
-// FIXME: Disabling test as qpp is the only backend capable
-// of resetting qubits right now.
-#if 0
 CUDAQ_TEST(NVQIRTester, checkQubitReset) {
   // Initialize two qubits in the 0-state.
   __quantum__rt__initialize(0, nullptr);
@@ -241,20 +246,21 @@ CUDAQ_TEST(NVQIRTester, checkQubitReset) {
       __quantum__rt__array_get_element_ptr_1d(qubits, 0));
   Qubit *q2 = *reinterpret_cast<Qubit **>(
       __quantum__rt__array_get_element_ptr_1d(qubits, 1));
-  
+
   // Place both qubits in the 1-state with X-gates.
   __quantum__qis__x(q1);
   __quantum__qis__x(q2);
   assert(*__quantum__qis__mz(q1) == 1);
   assert(*__quantum__qis__mz(q2) == 1);
 
-  // Reset just one of the qubits and confirm the other 
+  // Reset just one of the qubits and confirm the other
   // remains untouched.
   __quantum__qis__reset(q1);
   assert(*__quantum__qis__mz(q1) == 0);
   assert(*__quantum__qis__mz(q2) == 1);
+
+  __quantum__rt__qubit_release_array(qubits);
 }
-#endif
 
 Qubit *extract_qubit(Array *a, int idx) {
   auto q_raw_ptr = __quantum__rt__array_get_element_ptr_1d(a, idx);

```

Reply:
Thanks, I just added those and commented the resets back in elsewhere around this file. Just curious, any reason SWAP needs a `__quantum__rt__finalize` and reset doesn't?

Reply:
None of them actually need it - it has been left unimplemented, with the expectation we'll use it in the future (likely for MPI finalization)